Modules/_testcapi/heaptype.c

```diff
@@ -1008,6 +1008,24 @@ static PyType_Spec HeapCTypeSetattr_spec = {
     HeapCTypeSetattr_slots
 };
 
+static PyObject *heapctype_vectorcall(PyObject *self, PyObject *const *args_in,
+                                      size_t nargsf, PyObject *kwargs_in) {
+    return PyLong_FromLong(123);
+}
+
+static PyType_Slot HeapCTypeVectorcall_slots[] = {
+    {Py_tp_vectorcall, heapctype_vectorcall},
+    {0, 0},
+};
```

Comment:
That looks like a good example of how not to do it, since it leaves `tp_call` from the base's metaclass:
```pycon
>>> import _testcapi
>>> _testcapi.HeapCTypeVectorcall.__call__(_testcapi.HeapCTypeVectorcall, 'Sub', (), {})
<class '__main__.Sub'>
```
As per [the docs](https://docs.python.org/3/c-api/call.html#the-vectorcall-protocol), a class supporting vectorcall must also implement `tp_call` with the same semantics. Here, the overridden vectorcall function would need to match the behaviour of `PyType_Type->tp_call` (including any changes in future versions).

So, you do need to override `tp_new`/`tp_init`, and things get even more tricky if you allow subclasses of `HeapCTypeVectorcall`.

Reply:
I will override `tp_new` and make the type non-inheritable.

Reply:
I'm worried about leading users to a trap, and I'd prefer to see a full solution for the use case, to get an idea about where in the docs to put warnings, and what kinds of issues this opens.
Putting this in limited API means we should be extra careful about not exposing implementation details.

Reply:
I'm happy to implement a more representative version but wonder about the following: how should the vectorcall version communicate to the test suite that it was used? If the equivalence requirement you stated is interpreted strictly, it should not be possible to detect any difference.

For example, would it be OK I return two different instances from each mode of construction and then add a comment saying that these are only different for test-related purposes, and that the contract of these operations is that they should be interchangeable?

Reply:
Yes, that's fine for tests.